lib/plugin.c

```diff
@@ -168,7 +174,235 @@ static struct mptcpd_plugin_ops const *token_to_ops(mptcpd_token_t token)
 //                Plugin Registration and Management
 // ----------------------------------------------------------------
 
-bool mptcpd_plugin_load(char const *dir, char const *default_name)
+/**
+ * @struct plugin_info
+ *
+ * @brief Plugin information
+ */
+struct plugin_info
+{
+        /// Handle returned from call to @c dlopen().
+        void *handle;
+
+        /// Plugin descriptor.
+        struct mptcpd_plugin_desc const *desc;
+};
+
+/// List of @c plugin_info objects.
+static struct l_queue *_plugin_infos;
+
+/**
+ * @brief Compare plugin priorities.
+ *
+ * @param[in] a         New      plugin information.
+ * @param[in] b         Existing plugin information.
+ * @param[in] user_data User data (unused).
+ *
+ * @return Value greater than 0 if the new plugin should be inserted
+ *         after the existing plugin in the list, and a negative value
+ *         otherwise.
+ */
+static int compare_plugin_priority(void const *a,
+                                   void const *b,
+                                   void *user_data)
+{
+        (void) user_data;
+
+        struct plugin_info const *const new      = a;
+        struct plugin_info const *const existing = b;
+
+        /*
+          Cause "new" plugin to be inserted into the plugin list
+          before the "existing" plugin it its priority is higher
+          (lower value) than the existing plugin priority.
+        */
+        return new->desc->priority - existing->desc->priority;
+}
+
+static void report_error(int error, char const *msg)
+{
+        char errmsg[80];
+        int const r =
+                strerror_r(error, errmsg, L_ARRAY_SIZE(errmsg));
+
+        l_error("%s: %s", msg, r == 0 ? errmsg : "<unknown error>");
+}
+
+static void init_plugin(void *data, void *user_data)
+{
+        struct plugin_info const *const p  = data;
+        struct mptcpd_pm         *const pm = user_data;
+
+        if (p->desc->init && p->desc->init(pm) != 0)
+                l_warn("Plugin \"%s\" failed to initialize",
+                       p->desc->name);
+}
+
+static bool file_ends_in_so(char const *filename)
```

Comment:
Check out `l_str_has_suffix()`!

Reply:
Good catch.  I'll switch over to `l_str_has_suffix()`.  Thanks!